README.md
popclient
=========

This repository was forked from the "official" final resting place of the original 
`popclient` source code written by Carl Harris around 1993. `popclient` 
is a very small and simple POP3/POP2 email client that was developed 
in a time before POP and subsequently IMAP became the de-facto standards 
for e-mail retrieval on today's Internet. 

It is preserved here only for the sake of posterity, and almost certainly
shouldn't be used today. You might have read about `popclient` 
in Eric S Raymond's [The Cathedral and the Bazaar](http://www.catb.org/esr/writings/cathedral-bazaar/cathedral-bazaar/) or in Nick Christenson's 
[Sendmail Performance Tuning](https://books.google.com/books/about/Sendmail_Performance_Tuning.html) and perhaps have been curious to see (or even try to
build) the source. You might also be be interested in reviewing the original [man page](https://soulwing.github.io/popclient/popclient.html).

Around 1996 `popclient` was radically transformed by Eric Raymond into
the `fetchmail` utility. In 2004, [Rob Funk](https://sourceforge.net/u/robfunk/profile/) and others took responsibility for the ongoing maintenance and of
fetchmail, and as of 2013 it was still being actively maintained at
[www.fetchmail.info](http://www.fetchmail.info). If you're looking for 
command-line interface for POP, IMAP, etc, Fetchmail is for you.

Anyway, if you are trying to make it work, this patched version should work.
It has some enhancements, like the capability of using custom ports (for ssh forwarding)
or the option to use APOP authentication.


CHANGES

popclient
Copyright (c) 1993,1994  Carl Harris, Jr.  <ceharris@vt.edu>

Version	Description
------- -----------
1.0- 	o  Original POP2 development version, term support, minor patches.
1.6	

1.7	o  Added MAIL environment variable support, System V Makefile changes.

2.1	o  First version to support POP3.
	o  First wide release to public at large.
	o  Began maintaining this file of changes.

2.1a	o  Fixed bug in verbose output in POP3 mode.
	o  Changed bzero() on the options structure to account for programmer 
           boneheadedness.
	o  Changed errorneous call in SockPrintf from SockPuts to SockWrite.
	o  Moved error report on socket errors in main().
	o  Removed -g -O0 from CFLAGS in Makefile.

2.1b	o  Changed erroneous fputs() to fprintf() in POP3_OK.

2.20	o  Changed argv[] pointers to point to innocuous junk so that a
	   password specified using -p will not be compromised by the
	   'ps' command.
	o  Radically changed system mailbox handling to use /bin/mail (or
	   equivalent MDA) to deliver user's mail to the mail spool.  This
	   not only obviates the need for MAIL environment variable support,
	   but also relieves the client of folder locking concerns.  Thanks
	   go to Tor Lillqvist <tml@tik.vtt.fi> for this suggestion.
	o  Set up to allow installer to specify the default protocol and the
	   new MDA (/bin/mail) support configuration in config.h.
	o  Typecast call to getpass() to get rid of gcc warning.
	o  Check for INADDR_NONE definition in socket.h

2.21	o  Changed inline function argument definitions in socket.c to easy
	   SunOS compile.
	o  Fixed mail delivery to system mail folder to use _login_ ID not
	   POP mail user ID.


2.22
	o  Misc updates from spd@daphne.cps.unizar.es aka fsckawk
	o  Apr 24 1998: add the '-n port' option
	o  Oct 15 1998: add APOP support
	o  Nov  3 2017: Compatible with modern C compilers



pop2.c
/* Copyright 1993,1994 by Carl Harris, Jr.
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Carl Harris <ceharris@vt.edu>
 */


/***********************************************************************
  module:       pop2.c
  program:      popclient
  SCCS ID:	@(#)pop2.c	1.4  3/31/94
  programmer:   Carl Harris, ceharris@vt.edu
  date:         29 December 1993
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3 
  description:  POP2 client code.
 ***********************************************************************/
 
#include  <stdio.h>
#include  <sys/time.h>
#include  <unistd.h>
#include  <errno.h>

#include  "socket.h"
#include  "config.h"
#include  "popclient.h"

/* requisite, venerable SCCS ID */
static char sccs_id [] = "@(#)pop2.c	1.4\t3/31/94";

/* TCP port number for POP2 as defined by RFC 937 */
#define	  POP2_PORT	109

#ifndef NO_PROTO
/* prototypes for internal functions */
int POP2_sendcmd (char *cmd, int socket);
int POP2_sendHELO (char *userid, char *password, int socket);
int POP2_sendFOLD (char *folder, int socket);
int POP2_quit (int socket);
int POP2_stateGREET (int socket);
int POP2_stateNMBR (int socket);
int POP2_stateSIZE (int socket);
int POP2_stateXFER (int msgsize, int socket, int mboxfd, int topipe);
#endif


/*********************************************************************
  function:      doPOP2
  description:   retrieve messages from the specified mail server
                 using Post Office Protocol 2.

  arguments:     
    options      fully-specified options (i.e. parsed, defaults invoked,
                 etc).

  return value:  exit code from the set of PS_.* constants defined in 
                 popclient.h
  calls:         POP2_stateGREET, POP2_stateNMBR, POP2_stateSIZE,
                 POP2_stateXFER, POP2_sendcmd, POP2_sendHELO,
                 POP2_sendFOLD, POP2_quit, Socket, openuserfolder,
                 closeuserfolder, openmailpipe, closemailpipe.
  globals:       reads outlevel.
 *********************************************************************/

int doPOP2 (options)
struct optrec *options;
{
  int mboxfd;
  int socket;
  int number,msgsize,actsize;
  int status = PS_UNDEFINED;

  /* open the socket to the POP server */
  if ((socket = Socket(options->host,
	  options->port == 0 ? POP2_PORT : options->port
	  )) < 0) {
    perror("doPOP2: socket");
    return(PS_SOCKET);
  }
    
  /* open/lock the folder if it is a user folder or stdout */
  if (options->foldertype != OF_SYSMBOX)
    if ((mboxfd = openuserfolder(options)) < 0) 
      return(PS_FOLDER);
 
  /* wait for the POP2 greeting */
  if (POP2_stateGREET(socket) != 0) {
    POP2_quit(socket);
    return(PS_PROTOCOL);
  }

  /* log the user onto the server */
  POP2_sendHELO(options->userid,options->password,socket);
  if ((number = POP2_stateNMBR(socket)) < 0) {
    POP2_quit(socket);
    return(PS_AUTHFAIL);
  }

  /* set the remote folder if selected */
  if (*options->remotefolder != 0) {
    POP2_sendFOLD(options->remotefolder,socket);
    if ((number = POP2_stateNMBR(socket)) < 0) {
      POP2_quit(socket);
      return(PS_PROTOCOL);
    }
  }

  /* tell 'em how many messages are waiting */
  if (outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%d messages in folder %s\n",number,options->remotefolder);
  else
    ;

  /* fall into a retrieve/acknowledge loop */
  if (number > 0) { 

    POP2_sendcmd("READ",socket);
    msgsize = POP2_stateSIZE(socket);
    while (msgsize > 0) {

      /* open the pipe */
      if (options->foldertype == OF_SYSMBOX)
        if ((mboxfd = openmailpipe(options)) < 0) {   
          POP2_quit(socket);
          return(PS_FOLDER);
        }

      POP2_sendcmd("RETR",socket);
      actsize = POP2_stateXFER(msgsize,socket,mboxfd,
                               options->foldertype == OF_SYSMBOX);
      if (actsize == msgsize) 
        if (options->keep)
          POP2_sendcmd("ACKS",socket);
        else
          POP2_sendcmd("ACKD",socket);
      else if (actsize >= 0) 
        POP2_sendcmd("NACK",socket);
      else {
        POP2_quit(socket);
        return(PS_SOCKET);
      }

      /* close the pipe */
      if (options->foldertype == OF_SYSMBOX)
        if (closemailpipe(mboxfd) < 0) {
          POP2_quit(socket);
          return(PS_FOLDER);
        }
    
      msgsize = POP2_stateSIZE(socket);
    }
    POP2_quit(socket);
    status = msgsize == 0 ? PS_SUCCESS : PS_PROTOCOL;
  }
  else {
    POP2_quit(socket);
    status = PS_NOMAIL;
  }

  if (options->foldertype != OF_SYSMBOX)
    closeuserfolder(mboxfd);

  return(status);
}



/*********************************************************************
  function:      POP2_sendcmd
  description:   send a command string (with no arguments) a server.
  arguments:     
    cmd          command string to send.
    socket       socket to which the server is connected.

  return value:  none.
  calls:         SockPuts.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_sendcmd (cmd,socket) 
char *cmd;
int socket;
{
  SockPuts(socket,cmd);

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> %s\n",cmd);
  else
    ;
}


/*********************************************************************
  function:      POP2_sendHELO
  description:   send the HELO command to the server.
  arguments:     
    userid       user's mailserver id.
    password     user's mailserver password.
    socket       socket to which the server is connected.

  return value:  none.
  calls:         SockPrintf.
  globals:       read outlevel.
 *********************************************************************/

int POP2_sendHELO (userid,password,socket) 
char *userid, *password;
int socket;
{
  SockPrintf(socket,"HELO %s %s\r\n",userid,password);
    

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> HELO %s password\n",userid);
  else
    ;
}


/*********************************************************************
  function:      POP2_sendFOLD
  description:   send the FOLD command to the server.
  arguments:     
    folder       name of the folder to open on the server.
    socket       socket to which the server is connected.  

  return value:  none.
  calls:         SockPrintf.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_sendFOLD (folder,socket)
char *folder;
int socket;
{
  SockPrintf(socket,"FOLD %s\r\n",folder);

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> FOLD %s\n",folder);
  else
    ;
}


/*********************************************************************
  function:      POP2_quit
  description:   send the QUIT command to the server and close 
                 the socket.

  arguments:     
    socket       socket to which the server is connected.

  return value:  none.
  calls:         SockPuts.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_quit (socket)
int socket;
{
  SockPuts(socket,"QUIT");
  close(socket);

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> QUIT\n");
  else
    ;
}


/*********************************************************************
  function:      POP2_stateGREET
  description:   process the GREET state as described in RFC 937.
  arguments:     
    socket       ...to which server is connected.

  return value:  zero if server's handling of the GREET state was 
                 correct, else non-zero (may indicate difficulty
                 at the socket).
  calls:         SockGets.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_stateGREET (socket)
int socket;
{
  char buf [POPBUFSIZE];
 
  /* read the greeting from the server */
  if (SockGets(socket, buf, sizeof(buf)) == 0) {

    /* echo the server's greeting to the user */
    if (outlevel > O_SILENT)
      fprintf(stderr,"%s\n",buf);
    else
      ;
    /* is the greeting in the correct format? */
    if (*buf == '+')
      return(0);
    else
      return(-1);
  }
  else {
    /* an error at the socket */ 
    if (outlevel > O_SILENT)
      perror("error reading socket\n");
    else
      ;
    return(-1);
  }
}


/*********************************************************************
  function:      POP2_stateNMBR
  description:   process the NMBR state as described in RFC 937.
  arguments:     
    socket       ...to which the server is connected.

  return value:  zero if the expected NMBR state action occured, else
                 non-zero.  Following HELO, a non-zero return value 
                 usually here means the user authorization at the server
                 failed.
  calls:         SockGets.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_stateNMBR (socket)
int socket;
{
  int number;
  char buf [POPBUFSIZE];

  /* read the NMBR (#ccc) message from the server */
  if (SockGets(socket, buf, sizeof(buf)) == 0) {

    /* is the message in the proper format? */
    if (*buf == '#') {
      number = atoi(buf + 1);
      if (outlevel == O_VERBOSE)
        fprintf(stderr,"%s\n",buf);
      else
        ;
    }
    else {
      number = -1;
      if (outlevel > O_SILENT) 
        fprintf(stderr,"%s\n",buf);
      else
        ;
    }
  }
  else {
    /* socket problem */
    number = -1;
    if (outlevel == O_VERBOSE) 
      perror("socket read error\n");
    else
      ;
  }
  return(number);
}


/*********************************************************************
  function:      POP2_stateSIZE
  description:   process the SIZE state as described in RFC 937.
  arguments:     
    socket       ...to which the server is connected.

  return value:  zero if the expected SIZE state action occured, else
                 non-zero (usually indicates a protocol violation).
  calls:         SockGets.
  globals:       reads outlevel.
 *********************************************************************/

int POP2_stateSIZE (socket)
int socket;
{
  int msgsize;
  char buf [POPBUFSIZE];

  /* read the SIZE message (=ccc) from the server */
  if (SockGets(socket, buf, sizeof(buf)) == 0) 
    /* is the message in the correct format? */
    if (*buf == '=') {
      msgsize = atoi(buf + 1);
      if (outlevel == O_VERBOSE)
        fprintf(stderr,"%s\n",buf);
      else
        ;
    }
    else {
      msgsize = -1;
      if (outlevel > O_SILENT) 
        fprintf(stderr,"%s\n",buf);
      else
        ;
    }
  else {
    /* socket problem */
    msgsize = -1;
    if (outlevel == O_VERBOSE) 
      perror("socket read error\n");
    else
      ;
  }

  return(msgsize);
}


/*********************************************************************
  function:      POP2_stateXFER
  description:   process the XFER state as described in RFC 937.
  arguments:     
    msgsize      content length of the message as reported in the 
                 SIZE state.
    socket       ... to which the server is connected.
    mboxfd       open file descriptor to which the retrieved message will
                 be written.  
    topipe       true if we're writing to a the /bin/mail pipe.

  return value:  
    >= 0         actual length of the message received. 
    < 0          socket I/O problem.

  calls:         SockRead.
  globals:       reads outlevel. 
 *********************************************************************/

int POP2_stateXFER (msgsize,socket,mboxfd,topipe)
int msgsize;
int socket;
int mboxfd;
int topipe;
{
  int i,buflen,actsize;
  char buf [MSGBUFSIZE]; 
  time_t now;

  /* This keeps the retrieved message count for display purposes */
  static int msgnum = 0;  

  /* set up for status message if outlevel allows it */
  if (outlevel > O_SILENT && outlevel < O_VERBOSE) {
    fprintf(stderr,"reading message %d",++msgnum);
    /* won't do the '...' if retrieved messages are being sent to stdout */
    if (mboxfd == 1)  /* we're writing to stdout */
      fputs(".\n",stderr);
    else
      ;
  }
  else
    ;

  /* Unix mail folder format requires the Unix-syntax 'From' header.
     POP doesn't send it, so we fake it here. */
  if (!topipe) {
    now = time(NULL);
    sprintf(buf,"From POPmail %s",ctime(&now));
    if (write(mboxfd,buf,strlen(buf)) < 0) {
      perror("POP2_stateXFER: write");
      return(-1);
    }
  }

  /* read the specified message content length from the server */
  actsize = 0;
  while (msgsize > 0) {
    buflen = msgsize <= MSGBUFSIZE ? msgsize : MSGBUFSIZE;
    /* read a bufferful */ 
    if (SockRead(socket, buf, buflen) == 0) {
      /* write to folder, stripping CR chars in the process */
      for (i = 0;  i < buflen;  i++)
        if (*(buf + i) != '\r')
          if (write(mboxfd,buf + i,1) < 0) {
            perror("POP2_stateXFER: write");
            return(-1);
          }
          else
            ;  /* it was written */
        else
          ;  /* ignore CR character */
    }
    else
      return(-1);   /* socket problem */

    /* write another . for every bufferful received */
    if (outlevel > O_SILENT && outlevel < O_VERBOSE && mboxfd != 1) 
      fputc('.',stderr);
    else
      ;
    msgsize -= buflen;
    actsize += buflen;
  }

  if (!topipe) {
    /* The server may not write the extra newline required by the Unix
       mail folder format, so we write one here just in case */
    if (write(mboxfd,"\n",1) < 1) {
      perror("POP2_stateXFER: write");
      return(-1);
    }
  }
  else {
     /* the mailer might require some sort of termination string, send
        it if it is defined */
#ifdef BINMAIL_TERM
    if (write(mboxfd,BINMAIL_TERM,strlen(BINMAIL_TERM)) < 0) {
      perror("POP2_stateXFER: write");
      return(-1);
    }
#endif
  }

  /* finish up display output */
  if (outlevel == O_VERBOSE)
    fprintf(stderr,"(%d characters of message content)\n",actsize);
  else if (outlevel > O_SILENT && mboxfd != 0)
    fputc('\n',stderr);
  else
    ;

  return(actsize);
}


pop3.c
/* Copyright 1993,1994 by Carl Harris, Jr.
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Carl Harris <ceharris@vt.edu>
 */


/***********************************************************************
  module:       pop3.c
  program:      popclient
  SCCS ID:      @(#)pop3.c	2.4  3/31/94
  programmer:   Carl Harris, ceharris@vt.edu
  date:         29 December 1993
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3 
  description:  POP2 client code.
 ***********************************************************************/
 
#include  <stdio.h>
#include  <sys/time.h>
#include  <unistd.h>
#include  <ctype.h>
#include  <errno.h>

#include  "socket.h"
#include  "config.h"
#include  "popclient.h"

#include  "global.h"
#include  "md5.h"

/* requisite, venerable SCCS ID string */
static char sccs_id [] = "@(#)pop3.c	2.4\t3/31/94";

/* TCP port number for POP2 as defined by RFC 937 */
#define	  POP3_PORT	110

#define MD_CTX MD5_CTX
#define MDInit MD5Init
#define MDUpdate MD5Update
#define MDFinal MD5Final

#ifndef NO_PROTO
/* prototypes for internal functions */
int POP3_OK (char *buf, int socket);
int POP3_auth (char *userid, char *password, int socket);
int APOP_auth (char *userid, char *timestamp, char *password, int socket);
static void MDString (char *string);
int POP3_sendQUIT (int socket);
int POP3_sendSTAT (int *msgcount, int socket);
int POP3_sendRETR (int msgnum, int socket);
int POP3_sendDELE (int msgnum, int socket);
int POP3_readmsg (int socket, int mboxfd, int topipe);
#endif


/*********************************************************************
  function:      doPOP3
  description:   retrieve messages from the specified mail server
                 using Post Office Protocol 3.

  arguments:     
    options      fully-specified options (i.e. parsed, defaults invoked,
                 etc).

  return value:  exit code from the set of PS_.* constants defined in 
                 popclient.h
  calls:
  globals:       reads outlevel.
 *********************************************************************/

int doPOP3 (options)
struct optrec *options;
{
  int ok;
  int mboxfd;
  char buf [POPBUFSIZE];
  char *timestamp;
  int socket;
  int number,count;


  /* open the folder if we're not using the system mailbox */
  if (options->foldertype != OF_SYSMBOX) 
    if ((mboxfd = openuserfolder(options)) < 0) 
      return(PS_FOLDER);
    
  /* open the socket and get the greeting */
  if ((socket = Socket(options->host,
	  options->port == 0 ? POP3_PORT : options->port
	  )) < 0) {

    perror("doPOP3: socket");
    return(PS_SOCKET);
  }

  ok = POP3_OK(buf,socket);
  if (ok != 0) {
    if (ok != PS_SOCKET)
      POP3_sendQUIT(socket);
    close(socket);
    return(ok);
  }

  /* print the greeting */
  if (outlevel > O_SILENT && outlevel < O_VERBOSE) 
    fprintf(stderr,"%s\n",buf);
  else 
    ;

  /* spd@daphne.cps.unizar.es */
  /* get timestamp (for APOP) */

	if ( NULL == (timestamp=strchr(buf, '<')))
	{
		if ( options->whichpop == 4 )
		{
			fprintf(stderr,"Server can't do APOP\n",count);
			ok = PS_ERROR;
			goto cleanUp;
		}
	}


	if ( options->whichpop == 4 )
	{
		/* try to get authorized */
		ok = APOP_auth(options->userid,timestamp,options->password,socket);
		if (ok == PS_ERROR)
			ok = PS_AUTHFAIL;
		if (ok != 0)
			goto cleanUp;
	}
	else
	{
  /* try to get authorized */
  ok = POP3_auth(options->userid,options->password,socket);
  if (ok == PS_ERROR)
    ok = PS_AUTHFAIL;
  if (ok != 0)
    goto cleanUp;
	}

  /* find out how many messages are waiting */
  ok = POP3_sendSTAT(&count,socket);
  if (ok != 0) {
    goto cleanUp;
  }

  /* show them how many messages we'll be downloading */
  if (outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%d messages in folder\n",count);
  else
    ;

  if (count > 0) { 
    for (number = 1;  number <= count;  number++) {

      /* open the mail pipe if we're using the system mailbox */
      if (options->foldertype == OF_SYSMBOX) {
        ok = (mboxfd = openmailpipe(options)) < 0 ? -1 : 0;
        if (ok != 0)
          goto cleanUp;
      }
           
      ok = POP3_sendRETR(number,socket);
      if (ok != 0)
        goto cleanUp;
        
      ok = POP3_readmsg(socket,mboxfd,options->foldertype == OF_SYSMBOX);
      if (ok != 0)
        goto cleanUp;

      if (!options->keep) {
        ok = POP3_sendDELE(number,socket);
        if (ok != 0)
          goto cleanUp;
      }
      else
        ; /* message is kept */

      /* close the mail pipe if we're using the system mailbox */
      if (options->foldertype == OF_SYSMBOX) {
        ok = closemailpipe(mboxfd);
        if (ok != 0)
          goto cleanUp;
      }

    }
    ok = POP3_sendQUIT(socket);
    if (ok == 0)
      ok = PS_SUCCESS;
    close(socket);
    return(ok);
  }
  else {
    ok = POP3_sendQUIT(socket);
    if (ok == 0)
      ok = PS_NOMAIL;
    close(socket);
    return(ok);
  }

cleanUp:
  if (ok != 0 && ok != PS_SOCKET)
    POP3_sendQUIT(socket);

  if (options->foldertype != OF_SYSMBOX)
    if (closeuserfolder(mboxfd) < 0 && ok == 0)
      ok = PS_FOLDER;
    
  if (ok == PS_FOLDER || ok == PS_SOCKET) 
    perror("doPOP3: cleanUp");

  return(ok);
}



/*********************************************************************
  function:      POP3_OK
  description:   get the server's response to a command, and return
                 the extra arguments sent with the response.
  arguments:     
    argbuf       buffer to receive the argument string.
    socket       socket to which the server is connected.

  return value:  zero if okay, else return code.
  calls:         SockGets
  globals:       reads outlevel.
 *********************************************************************/

int POP3_OK (argbuf,socket)
char *argbuf;
int socket;
{
  int ok;
  char buf [POPBUFSIZE];
  char *bufp;

  if (SockGets(socket, buf, sizeof(buf)) == 0) {
    if (outlevel == O_VERBOSE)
      fprintf(stderr,"%s\n",buf);

    bufp = buf;
    if (*bufp == '+' || *bufp == '-')
      bufp++;
    else
      return(PS_PROTOCOL);

    while (isalpha(*bufp))
      bufp++;
    *(bufp++) = '\0';

    if (strcmp(buf,"+OK") == 0)
      ok = 0;
    else if (strcmp(buf,"-ERR") == 0)
      ok = PS_ERROR;
    else
      ok = PS_PROTOCOL;

    if (argbuf != NULL)
      strcpy(argbuf,bufp);
  }
  else 
    ok = PS_SOCKET;

  return(ok);
}



/*********************************************************************
  function:      POP3_auth
  description:   send the USER and PASS commands to the server, and
                 get the server's response.
  arguments:     
    userid       user's mailserver id.
    password     user's mailserver password.
    socket       socket to which the server is connected.

  return value:  non-zero if success, else zero.
  calls:         SockPrintf, POP3_OK.
  globals:       read outlevel.
 *********************************************************************/

int POP3_auth (userid,password,socket) 
char *userid, *password;
int socket;
{
  int ok;
  char buf [POPBUFSIZE];

  SockPrintf(socket,"USER %s\r\n",userid);
  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> USER %s\n",userid);

  if ((ok = POP3_OK(buf,socket)) == 0) {
    SockPrintf(socket,"PASS %s\r\n",password);
    if (outlevel == O_VERBOSE)
      fprintf(stderr,"> PASS password\n");

    if ((ok = POP3_OK(buf,socket)) == 0) 
      ;  /* okay, we're approved.. */
    else if (outlevel > O_SILENT && outlevel < O_VERBOSE)
      fprintf(stderr,"%s\n",buf);
    else
      ; /* say nothing */
  }
  else if (outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%s\n",buf);
  else
    ; /* say nothing */
  
  return(ok);
}



/*********************************************************************
  author:        spd@daphne.cps.unizar.es
  function:      APOP_auth
  description:   send the USER and PASS commands to the server, and
                 get the server's response.
  arguments:     
    userid       user's mailserver id.
    timestamp    rfc1939 APOP timestamp <process-ID.clock@hostname>
    password     user's mailserver password.
    socket       socket to which the server is connected.

  return value:  non-zero if success, else zero.
  calls:         SockPrintf, POP3_OK.
  globals:       read outlevel.
 *********************************************************************/

int APOP_auth (userid,timestamp,password,socket) 
char *userid, *timestamp, *password;
int socket;
{
  int ok;
  char buf [POPBUFSIZE];

  strncpy(buf, timestamp, sizeof(buf)-strlen(password));
  strcat(buf, password);

  MDString(buf);

  SockPrintf(socket,"APOP %s %s\r\n",userid, buf);
  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> USER %s\n",userid);

  if ((ok = POP3_OK(buf,socket)) == 0) 
    ;  /* okay, we're approved.. */
  else if (outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%s\n",buf);
  else
    ; /* say nothing */
  
  return(ok);
}
static void MDString (string)
char *string;
{
	register int i;
	char *p=string;
	MD_CTX context;
	unsigned char digest[16];
	unsigned int len = strlen (string);

	MDInit (&context);
	MDUpdate (&context, string, len);
	MDFinal (digest, &context);

	*string=0x00;
	
	for(i=0; i<16; i++)
	{
		sprintf( p, "%02x", digest[i]);
		p+=2;
	}
	*p=0x00;
}



/*********************************************************************
  function:      POP3_sendQUIT
  description:   send the QUIT command to the server and close 
                 the socket.

  arguments:     
    socket       socket to which the server is connected.

  return value:  none.
  calls:         SockPuts, POP3_OK.
  globals:       reads outlevel.
 *********************************************************************/

int POP3_sendQUIT (socket)
int socket;
{
  int ok;
  char buf [POPBUFSIZE];

  SockPuts(socket,"QUIT");

  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> QUIT\n");
  else
    ;

  ok = POP3_OK(buf,socket);
  if (ok != 0 && outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%s\n",buf);

  return(ok);
}



/*********************************************************************
  function:      POP3_sendSTAT
  description:   send the STAT command to the POP3 server to find
                 out how many messages are waiting.
  arguments:     
    count        pointer to an integer to receive the message count.
    socket       socket to which the POP3 server is connected.

  return value:  return code from POP3_OK.
  calls:         POP3_OK, SockPrintf
  globals:       reads outlevel.
 *********************************************************************/

int POP3_sendSTAT (msgcount,socket)
int *msgcount;
int socket;
{
  int ok;
  char buf [POPBUFSIZE];
  int totalsize;

  SockPrintf(socket,"STAT\r\n");
  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> STAT\n");
  
  ok = POP3_OK(buf,socket);
  if (ok == 0)
    sscanf(buf,"%d %d",msgcount,&totalsize);
  else if (outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%s\n",buf);

  return(ok);
}




/*********************************************************************
  function:      POP3_sendRETR
  description:   send the RETR command to the POP3 server.
  arguments:     
    msgnum       message ID number
    socket       socket to which the POP3 server is connected.

  return value:  return code from POP3_OK.
  calls:         POP3_OK, SockPrintf
  globals:       reads outlevel.
 *********************************************************************/

int POP3_sendRETR (msgnum,socket)
int msgnum;
int socket;
{
  int ok;
  char buf [POPBUFSIZE];

  SockPrintf(socket,"RETR %d\r\n",msgnum);
  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> RETR %d\n",msgnum);
  
  ok = POP3_OK(buf,socket);
  if (ok != 0 && outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%s\n",buf);

  return(ok);
}



/*********************************************************************
  function:      POP3_sendDELE
  description:   send the DELE command to the POP3 server.
  arguments:     
    msgnum       message ID number
    socket       socket to which the POP3 server is connected.

  return value:  return code from POP3_OK.
  calls:         POP3_OK, SockPrintF.
  globals:       reads outlevel.
 *********************************************************************/

int POP3_sendDELE (msgnum,socket)
int msgnum;
int socket;
{
  int ok;
  char buf [POPBUFSIZE];

  SockPrintf(socket,"DELE %d\r\n",msgnum);
  if (outlevel == O_VERBOSE)
    fprintf(stderr,"> DELE %d\n",msgnum);
  
  ok = POP3_OK(buf,socket);
  if (ok != 0 && outlevel > O_SILENT && outlevel < O_VERBOSE)
    fprintf(stderr,"%s\n",buf);

  return(ok);
}



/*********************************************************************
  function:      POP3_readmsg
  description:   Read the message content as described in RFC 1225.
  arguments:     
    socket       ... to which the server is connected.
    mboxfd       open file descriptor to which the retrieved message will
                 be written.  
    topipe       true if we're writing to the system mailbox pipe.

  return value:  zero if success else PS_* return code.
  calls:         SockGets.
  globals:       reads outlevel. 
 *********************************************************************/

int POP3_readmsg (socket,mboxfd,topipe)
int socket;
int mboxfd;
int topipe;
{
  char buf [MSGBUFSIZE];
  char *bufp;
  int lines,sizeticker;
  time_t now;
  /* This keeps the retrieved message count for display purposes */
  static int msgnum = 0;  

  /* set up for status message if outlevel allows it */
  if (outlevel > O_SILENT && outlevel < O_VERBOSE) {
    fprintf(stderr,"reading message %d",++msgnum);
    /* won't do the '...' if retrieved messages are being sent to stdout */
    if (mboxfd == 1)
      fputs(".\n",stderr);
    else
      ;
  }
  else
    ;

  /* Unix mail folder format requires the Unix-syntax 'From' header.
     POP3 doesn't send it, so we fake it here. */
  if (!topipe) {  
    now = time(NULL);
    sprintf(buf,"From POPmail %s",ctime(&now));
    if (write(mboxfd,buf,strlen(buf)) < 0) {
      perror("POP3_readmsg: write");
      return(PS_FOLDER);
    }
  }


  /* read the message content from the server */
  lines = 0;
  sizeticker = MSGBUFSIZE;
  while (1) {
    if (SockGets(socket,buf,sizeof(buf)) < 0)
      return(PS_SOCKET);
    bufp = buf;
    if (*bufp == '.') {
      bufp++;
      if (*bufp == 0)
        break;  /* end of message */
    }
    strcat(bufp,"\n");
    if (write(mboxfd,bufp,strlen(bufp)) < 0) {
      perror("POP3_readmsg: write");
      return(PS_FOLDER);
    }

    sizeticker -= strlen(bufp);
    if (sizeticker <= 0) {
      if (outlevel > O_SILENT && outlevel < O_VERBOSE)
        fputc('.',stderr);
      sizeticker = MSGBUFSIZE;
    }
    lines++;
  }

  if (!topipe) {
    /* The server may not write the extra newline required by the Unix
       mail folder format, so we write one here just in case */
    if (write(mboxfd,"\n",1) < 0) {
      perror("POP3_readmsg: write");
      return(PS_FOLDER);
    }
  }
  else {
    /* The mail delivery agent may require a terminator.  Write it if
       it has been defined */
#ifdef BINMAIL_TERM
    if (write(mboxfd,BINMAIL_TERM,strlen(BINMAIL_TERM)) < 0) {
      perror("POP3_readmsg: write");
      return(PS_FOLDER);
    }
#endif
    }

  /* finish up display output */
  if (outlevel == O_VERBOSE)
    fprintf(stderr,"(%d lines of message content)\n",lines);
  else if (outlevel > O_SILENT && mboxfd != 1) 
    fputs(".\n",stderr);
  else
    ;
  return(0);
}
